snps_accel_app: add iommu support

vadimbsnps opened the pull request:
This changes adds support for the IOMMU subsystem for the application driver devices created dynamically in the driver probe() function.

vadimbsnps committed:
snps_accel_app: add iommu support

This changes adds support for the IOMMU subsystem for the application
driver devices created dynamically in the driver probe() function.

pavelvkozlov commented on drivers/misc/snps_accel/snps_accel_mem.c:
```diff
@@ -17,7 +17,7 @@ snps_accel_mbuf_alloc(struct snps_accel_mem_ctx *mem, size_t size,
 	struct snps_accel_mem_buffer *mbuf = NULL;
 	struct snps_accel_file_priv *fpriv = to_snps_accel_file_priv(mem);
-	struct device *dmabuf_dev = mem->dev->parent;
+	struct device *dmabuf_dev = mem->dev;
```
Just to confirm — this change is necessary because the IOMMU setup is now being handled at the accelerator application device level (as I see in #25), rather than at the top-level accelerator device, correct?

vadimbsnps replied:
yes, correct.

pavelvkozlov commented on drivers/misc/snps_accel/snps_accel_drv.c:
```diff
@@ -503,6 +521,26 @@ static void snps_accel_release_app(struct snps_accel_app *accel_app)
+		domain = iommu_get_domain_for_dev(accel_app->device);
+		if (domain) {
+			iommu_detach_device(domain, accel_app->device);
+			iommu_domain_free(domain);
```
Do we really need to call ```iommu_domain_free()``` here? As we don’t explicitly call ```iommu_domain_alloc()``` and instead rely on ```of_iommu_configure()```, which attaches the device to the default IOMMU domain of the group. In this case, calling ```iommu_detach_device()``` should be sufficient. The IOMMU core should automatically release the default domain when its reference count drops to zero.
Also, it looks like we call ```iommu_group_get()``` but never call ```iommu_group_put()```, which might explain why we’re not seeing a double domain free in the IOMMU core.
Maybe it would be better to remove the ```iommu_domain_free()``` call and add ```iommu_group_put()``` just before the ```iommu_exit:``` label, to allow IOMMU core to release the domain?

vadimbsnps replied:
No, we don't. This causes a kernel panic in the driver unbind/bind sequence. I added a commit to this branch that removes this code. The presence of the iommu_detach_device() does not seem to have any difference, so I removed it too.

pavelvkozlov replied:
I'd keep the sequence: _get group -> get domain -> detach -> group_put_. It makes the flow clear to the reader (explicit detach after attach) and ensures the device is removed from the domain immediately. 
But I guess your solution is also acceptable, without detaching the device stays attached until the IOMMU core detects that the group is unused, which happens when ```device_destroy()``` is called, right?

vadimbsnps replied:
The driver does not call the `iommu_device_attach()` explicitly and it does not manage IOMMU domains manually. Instead it only calls the `of_iommu_configure()` which will call the `iommu_probe_device()` that takes care of the device registration with the IOMMU subsystem, automatically assigning it to an IOMMU group and domain. And when the driver calls `device_destroy()` in its `remove()` function the IOMMU subsystem will automatically unregister the device. So explicit IOMMU uninit code in the driver is not needed.

vadimbsnps committed:
snps_accel_add: fix device release from iommu

This fix solves the issue with the device unbind and bind sequence that
results in corrupted IOMMU device configuration.

pavelvkozlov reviewed (approved):
Ok, fine by me. Let's merge it
